### DISCUSSION TITLE
Compression of binary file using example code streamingHC_ringBuffer doesn't work

### DISCUSSION THREAD

**Discussion by vishalsingh11** [lz4/lz4#1365] [Q&A] [answered]
Hello @Cyan4973 

I tried your code (present in streamingHC_ringBuffer.c ) to compress binary file of 1 GB, but it is not working.

compressed file are bigger than uncompressed one.

Can you please help ?

**Comment by Cyan4973** [comment]
> it is not working.
> compressed file are bigger than uncompressed one.

So maybe the first question is to qualify the "not working" assessment:
is the code running and (trying to) compress the file, or not ?

> compressed file are bigger than uncompressed one.

How do you know the file is compressible ?
Have you tried to compress it with `lz4` directly ?

_edit_ : just for the record, I've tested `examples/streamingHC_ringBuffer` with a random compressible file (`enwik6`) and it compressed it fine.

**Reply by vishalsingh11** [reply]
Hello @Cyan4973 

Thank you for quick response. So following steps I followed.

1. compiled `examples/streamingHC_ringBuffer.c` code. 
2. generated a binary file of random 1 GB data with command -> `head -c 1G < /dev/urandom > myfile.bin`
3. invoked program with following command - `./decompressor myfile.bin` Note : I gave name of process `decompressor`.
4. I see following output - 

      ```
      input   = [myfile.bin]
      lz4     = [myfile.bin.lz4s-9]
      decoded = [myfile.bin.lz4s-9.dec]
      Verify : OK
      ``` 
      
 5. but when I check the details of compressed and uncompressed file, I see following (size of compressed file is 1.1G whereas original file is 1 G)
 
       ```
        -rwxrwxr-x 1 vishals domainusers  23K Feb 13 19:24 decompressor
        -rw-rw-r-- 1 vishals domainusers 5.8K Feb 13 19:24 Makefile
        -rw-rw-r-- 1 vishals domainusers 1.0G Feb 13 19:24 myfile.bin
        -rw-rw-r-- 1 vishals domainusers 1.1G Feb 13 19:25 myfile.bin.lz4s-9
        -rw-rw-r-- 1 vishals domainusers 1.0G Feb 13 19:25 myfile.bin.lz4s-9.dec
      ``` 
      
      result of md5 for reference (which shows that compressed and decompressed file is having same md5 value): 
      
      ```
      bb837593b5b3791e045e924f7501d7f4  myfile.bin
      41058954f7a4271034927522e03a88d5  myfile.bin.lz4s-9
      bb837593b5b3791e045e924f7501d7f4  myfile.bin.lz4s-9.dec
      ```

**Reply by Cyan4973** [reply]
> generated a binary file of random 1 GB data with command

Try using any (other) compression algorithm on this source.

**Reply by vishalsingh11** [reply]
Hello @Cyan4973 thank you for response
, so does it mean that binary files or images cannot be compressed using LZ4 ? 

I couldn't find out about this limitation of LZ4 anywhere !!! 

Can you please confirm? As this will help me in making decisions about the algorithm to choose.

**Reply by Cyan4973** [reply]
> so does it mean that binary files or images cannot be compressed using LZ4 ?

No, it means that random data cannot be compressed

> I couldn't find out about this limitation of LZ4 anywhere !!!

It has nothing to do with LZ4 in particular, it's a fundamental property of compression and even signal theory.

That's why I advised you to check if your sample was compressible, using any other compression algorithm.

**Reply by vishalsingh11** [reply]
But, I even tried to compress a valid ext4 image, and even that didn't worked.

```
input   = [test.ext4]
lz4     = [test.ext4.lz4s-9]
decoded = [test.ext4.lz4s-9.dec]
Verify : OK
``` 

compressed and decompressed file with size -

``` 
-rw-r--r-- 1 vishals domainusers 793M Feb 13 22:21 test.ext4
-rw-rw-r-- 1 vishals domainusers 804M Feb 13 22:22 test.ext4.lz4s-9
-rw-rw-r-- 1 vishals domainusers 793M Feb 13 22:22 test.ext4.lz4s-9.dec
``` 

md5sum value for reference : 


``` 
4f5fddda042fd9eecbfd89a0cc1b999c  test.ext4
a3986293fb6a3255013274bc6d996ad4  test.ext4.lz4s-9
4f5fddda042fd9eecbfd89a0cc1b999c  test.ext4.lz4s-9.dec
```

**Reply by Cyan4973** [reply, marked as answer]
> check if your sample was compressible, using any other compression algorithm.

**Reply by vishalsingh11** [reply]
alright, I got your point !!! 
Thank you so much @Cyan4973 and sorry for me being so silly.
